src/t_hash.c

```diff
@@ -2030,3 +2074,165 @@ void hrandfieldCommand(client *c) {
+
+    hashtable *ht = o->ptr;
+    void *entry_ptr = NULL;
+    int deleted = hashtablePop(ht, entry, &entry_ptr);
+
+    if (deleted) {
+        debugLogField(o, entry_ptr);
+        ctxAddEntry(ctx, entry_ptr);
+        server.stat_expiredfields++;
+        return 1;
+    }
+    return 0;
+}
+
+/* Free all entry memory for a given list of expired entries. */
+void freeEntries(void **entries, int n) {
+    for (int i = 0; i < n; i++) {
+        entryFree(entries[i]);
+    }
+}
+
+/* Free all robj references created in an argv list, except the shared command in slot 0. */
+void freeArgvObjects(robj **argv, int argc) {
+    for (int i = 1; i < argc; i++) {
+        // skip 1, the shared command
+        decrRefCount(argv[i]);
+    }
+}
+
+/* Build argv array for propagating HDEL of expired hash fields.
+ * argv[0] = shared.hdel
+ * argv[1] = key object
+ * argv[2..] = field names to delete
+ * Allocates a new key object, but *does not* free it — the caller must later.
+ * Returns the final argc.
+ */
+static int buildExpireFieldsArgv(void **entries, int n_entries, robj *o, robj *argv[]) {
+    int argc = 0;
+    robj *keyobj = createStringObjectFromSds(objectGetKey(o));
+    argv[argc++] = shared.hdel; // HDEL command
+    argv[argc++] = keyobj;      // key name
+    for (int i = 0; i < n_entries; i++) {
+        // field to delete
+        argv[argc++] = createStringObjectFromSds(entryGetField(entries[i]));
+    }
+    return argc;
+}
+
+/* Process expired fields for a hash key, deleting them,
+ * and propagating changes to replicas and AOF.
+ *
+ * This routine:
+ *  - identifies expired hash fields from a volatile set
+ *  - deletes them
+ *  - frees the entire key if the hash is empty
+ *  - propagates HDEL commands if the hash object isn't empty, and propagates hash DEL if empty.
+ *
+ * Returns the number of expired fields removed.
+ */
+size_t activeExpireFieldProcessKey(robj *o, serverDb *db, mstime_t now, unsigned long max_entries) {
+    /* Sanity check to prevent excessive stack allocation from large VLAs.
+     * We expect max_entries to be a small, bounded number (e.g. ~1000 max), which ~8k. */
+    serverAssert(max_entries > 0 && max_entries <= 1024);
+
+    /* skip TTL checks temporarily (to allow hashtable lookup) */
+    hashTypeIgnoreTTL(o, 1);
+    vset *vset = hashTypeGetVolatileSet(o);
+    serverAssert(!vsetIsEmpty(vset));
+
+    /* Temporary storage on stack for the entries and argv. */
+    void *entries[max_entries];
+    robj *argv[max_entries + 2]; // 2 extra slots for HDEL and key
+    int argc = 0;
+
+    expiryContext ctx = {.db = db, .key = o, .entries = entries, .n_entries = 0};
+
+    /* Pop expired fields from the volatile set. */
+    size_t expired = vsetPopExpired(vset, entryGetExpiry, expireEntry, now, max_entries, &ctx);
+
+    serverAssert(ctx.n_entries <= max_entries);
+
+    /* If no fields expired, restore normal TTL behavior and exit. */
+    if (expired == 0) {
+        hashTypeIgnoreTTL(o, 0);
```

Comment:
why don;t we just do that on the cleanup?

Reply:
yea that would work.

Reply:
O.K. I still feel it is safer and clear to do it there (you can add `if (o) hashTypeIgnoreTTL(o, 0);` there)